src/java/apiview-java-processor/src/main/java/com/azure/tools/apiview/processor/diagnostics/rules/BuilderTraitsDiagnosticRule.java

```diff
@@ -0,0 +1,137 @@
+        ));
+        traits.put("ConfigurationTrait", Arrays.asList(
+            new TraitMethod("configuration", "Configuration")
+        ));
+        traits.put("ConnectionStringTrait", Arrays.asList(
+            new TraitMethod("connectionString", "String")
+        ));
+        traits.put("EndpointTrait", Arrays.asList(
+            new TraitMethod("endpoint", "String")
+        ));
+        traits.put("HttpTrait", Arrays.asList(
+            new TraitMethod("httpClient", "HttpClient"),
+            new TraitMethod("pipeline", "HttpPipeline"),
+            new TraitMethod("addPolicy", "HttpPipelinePolicy"),
+            new TraitMethod("retryOptions", "RetryOptions"),
+            new TraitMethod("httpLogOptions", "HttpLogOptions"),
+            new TraitMethod("clientOptions", "ClientOptions")
+        ));
+        traits.put("AmqpTrait", Arrays.asList(
+            new TraitMethod("retryOptions", "AmqpRetryOptions"),
+            new TraitMethod("transportType", "AmqpTransportType"),
+            new TraitMethod("proxyOptions", "ProxyOptions"),
+            new TraitMethod("clientOptions", "ClientOptions")
+        ));
+    }
+
+    @Override
+    public void scanIndividual(final CompilationUnit cu, final APIListing listing) {
+        // If the CU has a @ServiceClientBuilder annotation:
+        //   Look for methods with names of methods we know should belong to the trait:
+        //     Ensure the matching trait is stated as being implemented. If the method exists but the trait interface is
+        //     not defined as being implemented, add a warning diagnostic that the trait should be implemented.
+        //     If the method name exists, but it has different parameters that what we expect, add a warning diagnostic
+        //     suggesting to use the trait.
+
+        cu.getTypes().forEach(type -> {
+            if (!type.isAnnotationPresent("ServiceClientBuilder")) return;
+
+            // iterate through trait-by-trait
+            for (Map.Entry<String, List<TraitMethod>> trait : traits.entrySet()) {
+                String traitName = trait.getKey();
+                List<TraitMethod> traitMethods = trait.getValue();
+
+                for (TraitMethod traitMethod : traitMethods) {
+                    String methodName = traitMethod.methodName;
+
+                    // firstly check if the exact method (name + params) is present
+                    List<MethodDeclaration> matchingMethods = type.getMethodsBySignature(methodName, traitMethod.methodParamTypes);
+                    if (!matchingMethods.isEmpty()) {
+                        // ensure we implement the trait
+                        if (!isTypeImplementingInterface(type, traitName)) {
+                            // we have an exact match to a trait method, but we do not implement the trait!
+                            for (MethodDeclaration m : matchingMethods) {
+                                listing.addDiagnostic(new Diagnostic(
```

Comment:
Should this wait until the end to make sure all methods in the trait exist on the client builder?

Reply:
In my testing the code appears to give the right amount of warnings. The design was such that we wanted to encourage people towards a trait, even if they are only part-way there.